### ISSUE TITLE
[Ubuntu (21.10) Linux] TModloader will not load for more than 1-3 seconds on Steam, then dies.

### ISSUE THREAD

**Issue by Trisscar05** [bluemagic123/tModLoader#2140]
### Version

Latest public-1.4-alpha

### OS

Linux

### Platform

Steam

### This bug affects

Gameplay as a Player

### Description

As the title, Tmod refuses to launch for more then the time it takes for Vulkan Shaders to apply, plus about 1-3 seconds once Steam is satisfied with the shaders. No logs are generated anywhere I have been able to find, including client-crashlog.txt
This being with no compatibility layers being forced.
With Proton 6 or 5 applied to both Tmod and Terraria, crashlogs are generated and popups for those crashes occur.

### Log File

1.3:
System.NotImplementedException: The method or operation is not implemented.
  at Terraria.ModLoader.Engine.HiDefGraphicsIssues.LogGraphicsDevice () [0x00010] in tModLoader.exe 
  at Terraria.ModLoader.Engine.HiDefGraphicsIssues.OnLoadContent () [0x00000] in tModLoader.exe 
  at Terraria.Main.LoadContent () [0x00000] in tModLoader.exe 
  at Microsoft.Xna.Framework.Game.Initialize () [0x0007f] in FNA.dll 
  at Terraria.Main.ClientInitialize () [0x00082] in tModLoader.exe 
  at Terraria.Main.Initialize () [0x064de] in tModLoader.exe 
  at Microsoft.Xna.Framework.Game.DoInitialize () [0x0003c] in FNA.dll 
  at Microsoft.Xna.Framework.Game.Run () [0x00011] in FNA.dll 
  at Terraria.Program.LaunchGame_ () [0x00082] in tModLoader.exe 

1.4:
System.IO.FileNotFoundException: Failed to find Native Library: Z:\home\jess\.steam\steam\steamapps\common\tModLoader\Libraries\Native\Windows\libc.dll
   at MonoLaunch.ResolveNativeLibrary(Assembly assembly, String name) in D:\a\tModLoader\tModLoader\src\tModLoader\Terraria\MonoLaunch.cs:line 86
   at System.Runtime.Loader.AssemblyLoadContext.GetResolvedUnmanagedDll(Assembly assembly, String unmanagedDllName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUnmanagedDllUsingEvent(String unmanagedDllName, Assembly assembly, IntPtr gchManagedAssemblyLoadContext)
   at MonoMod.RuntimeDetour.Platforms.DetourNativeLibcPlatform.mprotect(IntPtr start, IntPtr len, MmapProts prot)
   at MonoMod.RuntimeDetour.Platforms.DetourNativeLibcPlatform.SetMemPerms(IntPtr start, UInt64 len, MmapProts prot)
   at MonoMod.RuntimeDetour.Platforms.DetourNativeLibcPlatform.MakeWritable(IntPtr src, UInt32 size)
   at MonoMod.RuntimeDetour.DetourHelper.MakeWritable(IDetourNativePlatform plat, NativeDetourData detour)
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform._HookSelftest(MethodInfo from, MethodInfo to)
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform..ctor().........

### Reproduction frequency

Reliably and constantly.

### Additional Information

OS and graphics are updated, card is Nvidia GTX 1060 running driver 510.
It's been doing this for something like a week or two now. I've tried adding "export TERM=xterm" to both launch files (though Tmod didn't need it surprisingly), reinstalling everything, deleting the entire mods folder for Tmod, verifying files. Officially out of ideas.

**Comment by Solxanich** [comment]
There is currently a known issue coming from Steam that is breaking Linux.

Tmod 1.4 is intended to run native. Proton is unsupported. 
The current workaround, assuming a fresh installation is to run start-tmodloader.sh directly for it to complete architecture specific post installations. Once run once, Steam Launch can be used thereafter. 

If you have run with proton, it is recommended to delete the dotnet folder within the tmodloader install folder, and than run start-tmodloader.sh 

We are still investigating the issue. I've been informed it may be due to Steam trying to initialize its gui and interfering with the architectural deployment used in current scripts. We ask for your patience

Reference launch log for someone looking to fix:
[Launch_1.log](https://github.com/tModLoader/tModLoader/files/8056473/Launch_1.log)

**Comment by Trisscar05** [comment]
Valid and thanks much for the info, luck on your quest.

**Comment by Trisscar05** [comment]
That worked, I'm back in. Thanks much again!

**Comment by ghost** [comment]
> The current workaround, assuming a fresh installation is to run start-tmodloader.sh directly for it to complete architecture specific post installations. Once run once, Steam Launch can be used thereafter.

This is not the case on my end. I have to continue to run the game from a terminal.

No logs (or files for that matter) are being created in either `steamapps/common/tModLoader/tModLoader-Logs/` or `/home/helly/.local/share/Terraria/`

**Comment by JABorla** [comment]
This can be worked around by entering the settings for the launcher in steam, and under the Compatibility section, checking the box that says "Force the use of a specific Steam Play compatibility tool" and choosing "Steam Linux Runtime" under the dropdown of options.

**Comment by ghost** [comment]
### 2nd edit and **_fix_**:
After looking at the steam output when ran under a terminal, editing the run command to `%command% -mono` allows the game to launch as expected. No need to force steam runtime compatibility nor even run steam with the builtin runtime enabled.

The default of `%command%-mono` was the issue. The missing space was the problem.

I can only verify this on a current Arch system, not Ubuntu.


_----- old post ----_


Unfortunately that also doesn't work. I also tried using steam-native as well as steam-runtime.

Again, no logs are created/updated.

For what it's worth, this is on up-to-date (as of 5/3/2022) Arch.


Edit: 

Remembered you can run steam in a terminal to get errors games don't print to a log.

Forcing steam runtime compatibility:
```
/home/helly/.local/share/Steam/steamapps/common/SteamLinuxRuntime/apply-search-cwd.sh: 22: exec: /games/sam840p120/steamapps/common/tModLoader/start-tModLoader.sh-mono: not found
Game process removed: AppID 1281930 "/home/helly/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=1281930 -- '/games/sam840p120/steamapps/common/SteamLinuxRuntime_soldier'/_v2-entry-point --verb=waitforexitandrun -- '/home/helly/.local/share/Steam/steamapps/common/SteamLinuxRuntime'/scout-on-soldier-entry-point-v2 --  '/games/sam840p120/steamapps/common/tModLoader/start-tModLoader.sh'-mono", ProcID 10673
```

without:

```
/bin/sh\0-c\0/home/helly/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=1281930 -- '/games/sam840p120/steamapps/common/tModLoader/start-tModLoader.sh'-mono\0
Game process added : AppID 1281930 "/home/helly/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=1281930 -- '/games/sam840p120/steamapps/common/tModLoader/start-tModLoader.sh'-mono", ProcID 11115, IP 0.0.0.0:0
chdir /games/sam840p120/steamapps/common/tModLoader
pid 11116 != 11115, skipping destruction (fork without exec?)
Game process removed: AppID 1281930 "/home/helly/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=1281930 -- '/games/sam840p120/steamapps/common/tModLoader/start-tModLoader.sh'-mono", ProcID 11115
```

perhaps the issue is that there's no space between `'$dir/start-tModLoader.sh'` and `-mono`

**Comment by iSimpFor** [comment]
i believe i'm running into the same issue
[Launch.log](https://github.com/tModLoader/tModLoader/files/8696684/Launch.log)
[Natives.log](https://github.com/tModLoader/tModLoader/files/8696685/Natives.log)

**Comment by Chicken-Bones** [comment]
Fixed in #2380
